README.md
# env-config-map
[![Version][version-badge]][version-url]
[![Build][build-badge]][build-url]
[![Coverage][coverage-badge]][coverage-url]
[![Dependencies][dependencies-badge]][dependencies-url]
[![License][license-badge]][license-url]

Generates a config object from a config map using any input sources.  Mapping includes the most commonly encountered patterns such as setting default value, type casting, coercing null, coercing undefined, and redacting secrets for logging.  Default input source are environment variables from `process.env`.

- Zero dependency.
- Supported types:
  - `string`
  - `number`
  - `boolean`
  - `object`
  - `arrayCommaDelim`
  - _also supports custom types_
- `redact` option to redact value for logging.
- `coerceNull` option to coerce `"null"` to `null`.
- `coerceUndefined` option to coerce `"undefined"` to `undefined`.
- `getter` option to get input values from other sources. Default source is `process.env`.

## Installation
```console
npm install env-config-map
```

## Run Sandbox Example
```console
npm run sandbox
```

## Sandbox Example
```js
// source from .env (optional)
// require('dotenv').config();

const envConfigMap = require('env-config-map');

// setup fixture data for example
process.env.SERVER_HOST = '0.0.0.0';
process.env.SERVER_PORT = 8080;
process.env.MAINTENANCE_MODE = 'true';
process.env.ENABLE_PROFILER = 'NO';
process.env.SETTINGS = '{ "path": "/tmp", "timeout": 1000 } ';
process.env.ACCESS_KEY = 'myAccessKey';
process.env.COERCE_NULL_DEFAULT = 'null';
process.env.COERCE_NULL_DISABLED = 'null';
// process.env.LOG_LEVEL = undefined;

const configMap = {
  SERVER_HOST: { default: 'localhost' },
  SERVER_PORT: { default: 80, type: 'number' },
  MAINTENANCE_MODE: { default: false, type: 'boolean' },
  ENABLE_PROFILER: { default: false, type: 'booleanYesNo' },
  SETTINGS: { type: 'object' },
  ACCESS_KEY: { redact: true },
  COERCE_NULL_DEFAULT: {},
  COERCE_NULL_DISABLED: { coerceNull: false },
  LOG_LEVEL: { default: 'info' },
};

const options = {
  types: {
    // define custom type "booleanYesNo"
    booleanYesNo: (str) => {
      const normalized = envConfigMap.utils.lowerTrim(str);
      if (normalized === 'yes') return true;
      if (normalized === 'no') return false;
      return null;
    },
  },
  // customize redactor
  redactor: (str) => str.replace(/.+/, '*** REDACTED ***'),
};

const config = envConfigMap(configMap, options);

console.log(config);
console.log(config.getRedacted());
```

### console.log(config);
```js
{
  SERVER_HOST: '0.0.0.0',
  SERVER_PORT: 8080,
  MAINTENANCE_MODE: true,
  ENABLE_PROFILER: false,
  SETTINGS: { path: '/tmp', timeout: 1000 },
  ACCESS_KEY: 'myAccessKey',
  COERCE_NULL_DEFAULT: null,
  COERCE_NULL_DISABLED: 'null',
  LOG_LEVEL: 'info',
  getRedacted: [Function],
  getOptions: [Function]
}
```

### console.log(config.getRedacted());
```js
{
  SERVER_HOST: '0.0.0.0',
  SERVER_PORT: 8080,
  MAINTENANCE_MODE: true,
  ENABLE_PROFILER: false,
  SETTINGS: { path: '/tmp', timeout: 1000 },
  ACCESS_KEY: '*** REDACTED ***',
  COERCE_NULL_DEFAULT: null,
  COERCE_NULL_DISABLED: 'null',
  LOG_LEVEL: 'info'
}
```

## configMap
- `default` : _mixed_
  - Sets the default value.
- `type` : _string_ _(default: string)_
  - Specify the type for the key. Cast operation will call the type casting function defined in `options.types`.
    - `string`
    - `number`
    - `boolean`
    - `object`
    - `arrayCommaDelim`
- `redact` : _boolean_
  - Flag to indicate if the value is a secret and needs to be redacted.
- `coerceNull` : _boolean_
  - Coerce string `'null'` to `null`. Supersedes `options.coerceNull`.
- `coerceUndefined` : _boolean_
  - Coerce string `'undefined'` to `undefined`. Supersedes `options.coerceNull`.
```js
const configMap = {
  SERVER_PORT: { default: 80, type: 'number' },
  ACCESS_KEY: { redact: true },
  ENABLE_PROFILER: { default: false, type: 'booleanYesNo' },
  COERCE_DISABLED: { coerceNull: false, coerceUndefined: false },
};
const config = envConfigMap(configMap, options);
```

## options
- `getter` : _function_
  - Get input value for key. Default source is `process.env`.
- `types` : _object_
  - For defining additional types and it will merge with the supported types.
- `redactor` : _function_
  - Function to redact value flaged with the `redact` configMap option.
- `coerceNull` : _boolean_ _(default: true)_
  - Coerce string `'null'` to `null`.
- `coerceUndefined` : _boolean_ _(default: true)_
  - Coerce string `'undefined'` to `undefined`.
```js
const options = {
  getter: (key) => process.env[key],
  types: {
    // define custom type "booleanYesNo"
    booleanYesNo: (str) => {
      const normalized = envConfigMap.utils.lowerTrim(str);
      if (normalized === 'yes') return true;
      if (normalized === 'no') return false;
      return null;
    },
  },
  // customized redactor
  redactor: (str) => str.replace(/.+/, '*** REDACTED ***'),
  coerceNull: true,
  coerceUndefined: false,
};
const config = envConfigMap(configMap, options);
```

## Misc Exports
```js
const envConfigMap = require('env-config-map');
const defaultOptions = envConfigMap.defaultOptions;
const supportedTypes = envConfigMap.types;
const helperUtils = envConfigMap.utils;
```

## Example Server Config with `env-config-map`
### .env
```js
APP_NAME = envConfigMap;
SERVER_PORT = 8080;
ACCESS_KEY = myAccessKey;
```

### config.js
```js
require('dotenv').config();
const envConfigMap = require('env-config-map');

const configMap = {
  APP_NAME: { default: 'noname' },
  SERVER_HOST: { default: 'localhost' },
  SERVER_PORT: { default: 80, type: 'number' },
  ACCESS_KEY: { redact: true },
};
const config = envConfigMap(configMap);

module.exports = config;
```

### server.js
```js
const http = require('http');
const config = require('./config.js');

console.log('Configs loaded.', config.getRedacted());

http
  .createServer((req, res) => {
    res.write(`Hello ${config.APP_NAME}!`);
    res.end();
  })
  .listen(config.SERVER_PORT, config.SERVER_HOST, () => console.log(
    `server listening on ${config.SERVER_HOST}:${config.SERVER_PORT}`,
  ));
```

## License
Copyright &copy; 2019 Joe Yu. Licensed under the [MIT License][license-url].


[version-badge]: https://img.shields.io/npm/v/env-config-map.svg?style=flat-square
[version-url]: https://www.npmjs.com/package/env-config-map

[build-badge]: https://img.shields.io/circleci/build/github/hexkode/env-config-map/master.svg?style=flat-square
[build-url]: https://circleci.com/gh/hexkode/env-config-map/tree/master

[coverage-badge]: https://img.shields.io/coveralls/github/hexkode/env-config-map.svg?style=flat-square
[coverage-url]: https://coveralls.io/github/hexkode/env-config-map?branch=master

[dependencies-badge]: https://img.shields.io/david/hexkode/env-config-map.svg?style=flat-square
[dependencies-url]: https://david-dm.org/hexkode/env-config-map

[license-badge]: https://img.shields.io/npm/l/env-config-map.svg?style=flat-square
[license-url]: LICENSE


CHANGELOG.md
# Changelog

## 1.0.4
###### 2019-11-08
- Bumps eslint-utils from 1.4.0 to 1.4.3.

## 1.0.3
###### 2019-08-12
- Tweaked .eslintrc.json and ran lint fix.
- Updated README.md.

## 1.0.2
###### 2019-08-12
- Updated README.md.

## 1.0.1
###### 2019-08-11
- Added CHANGELOG.md.
- Added License section in README.md.
- Updated badges in README.md.

## 1.0.0
###### 2019-08-11
- First stable release.


test/index.test.js
const envConfigMap = require('../index.js');

test('baseline test equal expected output', () => {
  process.env.NODE_ENV = 'test';
  process.env.SERVER_PORT = '8080';
  process.env.ENABLE_CORS = 'true';
  process.env.DB_PASSWORD = 'mypassword';
  process.env.DB_ENABLE_PROFILER = 'yes';
  process.env.EXAMPLE_OBJECT = '{ "retry": 3, "timeout": 1000 } ';
  process.env.EXAMPLE_ARRAY = '[ "a", 1 ]';
  process.env.EXAMPLE_ARRAY_COMMA_DELIM = 'id,email,   dateCreated   ,dateModified';
  process.env.NOT_IN_CONFIG_MAP = 'not mapped';
  process.env.MISSING_TYPE_TRANSFORM = 'pass thru          ';

  const configMap = {
    NODE_ENV: { default: 'development' },
    LOG_LEVEL: { default: 'info' },
    SERVER_HOST: {},
    SERVER_PORT: { default: 80, type: 'number' },
    ENABLE_CORS: { default: false, type: 'boolean' },
    DB_PASSWORD: { isSecret: true },
    DB_ENABLE_PROFILER: { default: false, type: 'booleanYesNo' },
    EXAMPLE_OBJECT: { type: 'object' },
    EXAMPLE_ARRAY: { type: 'object' },
    EXAMPLE_ARRAY_COMMA_DELIM: { type: 'arrayCommaDelim' },
    MISSING_TYPE_TRANSFORM: { type: 'typeNotDefined' },
  };

  const options = {
    redactedString: 'XXXXXXXXXX',
    typeTransform: {
      booleanYesNo: string => (string === 'yes' ? true : false),
    },
  };

  const config = envConfigMap(configMap, options);

  expect(config).toMatchSnapshot();
  expect(config.getRedacted()).toMatchSnapshot();
  expect(config.getDefaultOptions()).toMatchSnapshot();
  expect(config.getOptions()).toMatchSnapshot();
});

test('getOptions() should be equal to default options when no options are provided', () => {
  const config = envConfigMap({});

  expect(config.getOptions()).toStrictEqual(config.getDefaultOptions());
});


test/unit/index.test.js
const envConfigMap = require('../../index.js');

describe('default redaction', () => {
  test('should equal to "**********"', () => {
    process.env.FIXTURE = 'mypassword';
    const configMap = {
      FIXTURE: { redact: true },
    };

    const config = envConfigMap(configMap);
    const redacted = config.getRedacted();
    expect(redacted.FIXTURE).toStrictEqual('**********');
  });
});


test/unit/types.test.js
const types = require('../../src/types.js');

describe('types', () => {
  describe('string', () => {
    it('returns input', () => {
      expect(types.string('')).toStrictEqual('');
      expect(types.string('   ')).toStrictEqual('   ');
      expect(types.string('   testValuePadded   ')).toStrictEqual('   testValuePadded   ');
      expect(types.string('0')).toStrictEqual('0');
      expect(types.string('3.14')).toStrictEqual('3.14');
      expect(types.string('12345')).toStrictEqual('12345');
      expect(types.string('null')).toStrictEqual('null');
      expect(types.string('undefined')).toStrictEqual('undefined');
    });
  });

  describe('number', () => {
    it('returns integer for integer string', () => {
      expect(types.number('0')).toStrictEqual(0);
      expect(types.number('3')).toStrictEqual(3);
    });

    it('returns float for float string', () => {
      expect(types.number('3.14')).toStrictEqual(3.14);
      expect(types.number('0.009')).toStrictEqual(0.009);
    });

    it('returns null for strings that are not numeric', () => {
      expect(types.number('invalidNumberString')).toBeNull();
      expect(types.number('null')).toBeNull();
      expect(types.number('undefined')).toBeNull();
    });
  });

  describe('boolean', () => {
    it('returns true for case insensitive "true" and "1"', () => {
      expect(types.boolean('true')).toStrictEqual(true);
      expect(types.boolean('1')).toStrictEqual(true);
      expect(types.boolean(' TrUe   ')).toStrictEqual(true);
    });

    it('returns false for case insensitive "false" and "0"', () => {
      expect(types.boolean('false')).toStrictEqual(false);
      expect(types.boolean('0')).toStrictEqual(false);
      expect(types.boolean('FALSE')).toStrictEqual(false);
    });

    it('returns null for strings that are not boolean like', () => {
      expect(types.boolean('neTrue1False0')).toBeNull();
      expect(types.boolean('null')).toBeNull();
      expect(types.boolean('undefined')).toBeNull();
    });
  });

  describe('object', () => {
    const objectFixture = {
      a: 1,
      b: 'str1',
      c: [],
      d: true,
      e: {
        f: [2, 3, 'str2'],
      },
    };

    it('returns object for valid stringify object', () => {
      expect(types.object('{ "a": 1, "b": "str1", "c": [], "d": true, "e": { "f": [ 2 , 3, "str2"] } }')).toStrictEqual(
        objectFixture,
      );
    });

    it('throws exception for invalid stringify object', () => {
      expect(() => types.object('{ "a": 1')).toThrow();
      expect(() => types.object('undefined')).toThrow();
    });

    it('returns null for "null"', () => {
      expect(types.object('null')).toBeNull();
    });
  });

  describe('arrayCommaDelim', () => {
    it('returns array for comma delimited string', () => {
      expect(types.arrayCommaDelim('a,b,c')).toStrictEqual(['a', 'b', 'c']);
    });

    it('returns array with all items trimmed', () => {
      expect(types.arrayCommaDelim(' a ,  b   , c ')).toStrictEqual(['a', 'b', 'c']);
    });

    it('returns array with length of 1 for any one word string', () => {
      expect(types.arrayCommaDelim('null')).toStrictEqual(['null']);
      expect(types.arrayCommaDelim('undefined')).toStrictEqual(['undefined']);
    });
  });
});


src/index.js
const types = require('./types.js');
const utils = require('./utils.js');

const defaultOptions = {
  types,
  getter: utils.getter,
  redactor: utils.redactor,
  coerceUndefined: true,
  coerceNull: true,
};

/**
 * Maps values returned from opts.getter() to a config object.
 * Also generates a redacted version of the config object for logging.
 *
 * @param {Object} configMap
 * @param {Object} options
 * @returns {Object}
 */
const envConfigMap = (configMap = {}, options = {}) => {
  const opts = {
    ...defaultOptions,
    ...options,
    types: {
      ...defaultOptions.types,
      ...options.types,
    },
  };

  const config = {};
  const redacted = {};

  config.getRedacted = () => redacted;
  config.getOptions = () => opts;

  Object.keys(configMap).forEach((key) => {
    const props = configMap[key];
    const value = opts.getter(key);
    const type = props.type || 'string';
    const cast = opts.types[type];
    const coerceUndefined = typeof props.coerceUndefined === 'boolean' ? props.coerceUndefined : opts.coerceUndefined;
    const coerceNull = typeof props.coerceNull === 'boolean' ? props.coerceNull : opts.coerceNull;

    let converted;
    if (value === undefined) {
      converted = props.default;
    } else {
      converted = utils.convert(value, cast, { coerceUndefined, coerceNull });
    }

    config[key] = converted;
    redacted[key] = props.redact === true && converted ? opts.redactor(converted) : converted;
  });

  return config;
};

module.exports = envConfigMap;
module.exports.defaultOptions = defaultOptions;
module.exports.types = types;
module.exports.utils = utils;


src/utils.js
const DEFAULT_REDACTED = '**********';

/**
 * Function to generate redacted values.
 *
 * @param {string} str
 * @returns {string}
 */
const redactor = () => DEFAULT_REDACTED;

/**
 * Function to source config values for mapping.
 *
 * @param {string} key
 * @returns {*}
 */
const getter = (key) => process.env[key];

/**
 * If input is string type, it applies lower case and trim,
 * anything else returns empty string.
 *
 * @param {*} str
 * @returns {string}
 */
const lowerTrim = (str) => (typeof str === 'string' ? str.toLowerCase().trim() : '');

/**
 * Coerce undefined from string
 *
 * @param {string} str
 * @returns {boolean}
 */
const coerceUndefinedString = (str) => lowerTrim(str) === 'undefined';

/**
 * Coerce null from string
 *
 * @param {string} str
 * @returns {boolean}
 */
const coerceNullString = (str) => lowerTrim(str) === 'null';

/**
 * Will only call cast() with a string type value.
 * Any other types will be filtered and returned as undefined or null.
 * If cast is not a function, input will be returned as is.
 *
 * @param {*} mixed
 * @param {function} cast
 * @param {Object} options
 * @returns {*}
 */
const convert = (mixed, cast, options = {}) => {
  const { coerceUndefined = true, coerceNull = true } = options;

  // string type filter
  if (typeof mixed !== 'string') return mixed === undefined ? undefined : null;

  // coerce undefined filter
  if (coerceUndefined === true && coerceUndefinedString(mixed) === true) return undefined;

  // coerce null filter
  if (coerceNull === true && coerceNullString(mixed) === true) return null;

  // validate cast function filter
  if (typeof cast !== 'function') return mixed;

  return cast(mixed);
};

module.exports = {
  DEFAULT_REDACTED,
  redactor,
  getter,
  convert,
  coerceUndefinedString,
  coerceNullString,
  lowerTrim,
};


src/typeTransform.js
const packageJson = require('../package.json');

const string = string => {
  if (typeof string === 'string') return string.trim();
  return _normalize(string);
};

const number = string => {
  if (typeof string === 'number') return string;

  const normalized = _normalize(string);
  if (!normalized) return normalized;

  const number = Number(normalized);
  return Number.isNaN(number) ? undefined : number;
};

const boolean = string => {
  if (typeof string === 'boolean') return string;

  const normalized = _normalize(string);
  if (!normalized) return normalized;

  if (normalized === 'true' || normalized === '1') {
    return true;
  } else if (normalized === 'false' || normalized === '0') {
    return false;
  } else {
    return undefined;
  }
};

const object = string => {
  if (typeof string === 'object') return string;

  const normalized = _normalize(string);
  if (!normalized) return normalized;

  let results = undefined;
  try {
    results = JSON.parse(normalized);
  } catch (err) {
    // console.error(`[${packageJson.name}] Exception encountered when converting from stringify json to object.`, err);
  }

  return results;
};

const arrayCommaDelim = string => {
  const normalized = _normalize(string);
  if (!normalized) return normalized;

  const results = normalized.split(',');
  return results.map(result => result.trim());
};

/**
 * normalize any type to non-empty trimmed string before transform.
 * undefined and null are passed thru.
 *
 * @param {*} value
 * @param {string|null|undefined} options
 */
const _normalize = value => {
  let normalized = undefined;

  if (value === undefined || value === null) {
    return value;
  }

  if (typeof value === 'string') {
    const trimmed = value.trim();
    if (trimmed) {
      normalized = trimmed;
    }
  }

  return normalized;
};

const _undefined = value => {
  if (typeof value === 'string') {
    const trimmed = value.trim();
    if (trimmed === 'undefined') {
      return undefined;
    }
  }

  return value;
};

const _null = value => {
  if (typeof value === 'string') {
    const trimmed = value.trim();
    if (trimmed === 'null') {
      return null;
    }
  }

  return value;
};

module.exports = {
  string,
  number,
  boolean,
  object,
  arrayCommaDelim,
  _normalize,
  _undefined,
  _null,
};


sandbox.js
/* eslint-disable no-console */
// source from .env (optional)
// require('dotenv').config();

const envConfigMap = require('./index');

// setup fixture data for example
process.env.SERVER_HOST = '0.0.0.0';
process.env.SERVER_PORT = 8080;
process.env.MAINTENANCE_MODE = 'true';
process.env.ENABLE_PROFILER = 'NO';
process.env.SETTINGS = '{ "path": "/tmp", "timeout": 1000 } ';
process.env.ACCESS_KEY = 'myAccessKey';
process.env.COERCE_NULL_DEFAULT = 'null';
process.env.COERCE_NULL_DISABLED = 'null';
// process.env.LOG_LEVEL = undefined;

const configMap = {
  SERVER_HOST: { default: 'localhost' },
  SERVER_PORT: { default: 80, type: 'number' },
  MAINTENANCE_MODE: { default: false, type: 'boolean' },
  ENABLE_PROFILER: { default: false, type: 'booleanYesNo' },
  SETTINGS: { type: 'object' },
  ACCESS_KEY: { redact: true },
  COERCE_NULL_DEFAULT: {},
  COERCE_NULL_DISABLED: { coerceNull: false },
  LOG_LEVEL: { default: 'info' },
};

const options = {
  types: {
    // define custom type "booleanYesNo"
    booleanYesNo: (str) => {
      const normalized = envConfigMap.utils.lowerTrim(str);
      if (normalized === 'yes') return true;
      if (normalized === 'no') return false;
      return null;
    },
  },
  // customize redactor
  redactor: (str) => str.replace(/.+/, '*** REDACTED ***'),
};

const config = envConfigMap(configMap, options);

console.log(config);
console.log(config.getRedacted());
